File: packages/go/echozero/hmac_test.go
package echozero

import "testing"

func TestSignRestRequestMatchesBackendPayloadFormat(t *testing.T) {
	headers, err := SignRestRequest(
		"test_secret",
		"POST",
		"/api/api-keys",
		map[string]any{"name": "SDK HMAC Test"},
		1_710_000_000_000,
	)
	if err != nil {
		t.Fatal(err)
	}

	const expected = "274c9ff280eadf751530e9e7fce2c2a573d8676b13b7108fe353407df7cc9e00"
	if headers.Signature != expected {
		t.Fatalf("signature mismatch: got %s want %s", headers.Signature, expected)
	}
}

func TestSignInboundWebhookUsesCanonicalJSON(t *testing.T) {
	headers, err := SignInboundWebhook(
		"test_secret",
		map[string]any{
			"text":           "BUY SOL 500 USDC",
			"idempotencyKey": "sdk-test-1",
		},
		1_710_000_000,
	)
	if err != nil {
		t.Fatal(err)
	}

	const expected = "1d30d896fc609e62bcf9be991c1dc1177a9c63219909869ef2846bad4685de3b"
	if headers.Signature != expected {
		t.Fatalf("signature mismatch: got %s want %s", headers.Signature, expected)
	}
}

func TestInboundCanonicalStripsUnknownFields(t *testing.T) {
	canonical, err := InboundWebhookCanonicalJSON(map[string]any{
		"text":       "BUY SOL",
		"extraField": "ignored",
	})
	if err != nil {
		t.Fatal(err)
	}
	if canonical != `{"text":"BUY SOL"}` {
		t.Fatalf("unexpected canonical: %s", canonical)
	}
}

func TestSignInboundWebhookIgnoresUnknownFields(t *testing.T) {
	body := map[string]any{
		"eventType":      "buy",
		"idempotencyKey": "test-1",
		"reasoning":      "test",
		"tokenAddress":   "So11111111111111111111111111111111111111112",
		"amount":         500,
	}
	withExtra := map[string]any{
		"eventType":      "buy",
		"idempotencyKey": "test-1",
		"reasoning":      "test",
		"tokenAddress":   "So11111111111111111111111111111111111111112",
		"amount":         500,
		"unknownField":   "strip me",
	}
	a, err := SignInboundWebhook("secret", body, 1_710_000_000)
	if err != nil {
		t.Fatal(err)
	}
	b, err := SignInboundWebhook("secret", withExtra, 1_710_000_000)
	if err != nil {
		t.Fatal(err)
	}
	if a.Signature != b.Signature {
		t.Fatalf("expected equal signatures, got %s vs %s", a.Signature, b.Signature)
	}
	const expected = "be420f61d91e6b871481774c62f972f0aafa6f5f1a727ba1e4a32558784f77c3"
	if a.Signature != expected {
		t.Fatalf("signature mismatch: got %s want %s", a.Signature, expected)
	}
}

func TestVerifyInboundWebhook(t *testing.T) {
	body := map[string]any{
		"text":           "BUY SOL 500 USDC",
		"idempotencyKey": "sdk-test-1",
	}
	headers, err := SignInboundWebhook("test_secret", body, 1_710_000_000)
	if err != nil {
		t.Fatal(err)
	}

	ok, err := VerifyInboundWebhook("test_secret", body, 1_710_000_000, headers.Signature, 999_999_999)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected webhook verification to pass")
	}
}

func TestVerifyOutboundWebhook(t *testing.T) {
	rawBody := `{"event":"signal.execution","signalId":"sig_1","developerAgentId":"agent_1","status":"executed","timestamp":"2026-07-06T12:00:00.000Z"}`
	timestamp := "2026-07-06T12:00:00.000Z"
	signature := hmacSHA256Hex("webhook_secret", timestamp+"."+rawBody)
	if !VerifyOutboundWebhook("webhook_secret", rawBody, timestamp, signature) {
		t.Fatal("expected outbound webhook verification to pass")
	}
}

File: packages/typescript/src/hmac.test.ts
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';
import { inboundWebhookCanonicalJson } from './inboundCanonical.js';
import {
  signInboundWebhook,
  signRestRequest,
  verifyInboundWebhook,
  verifyOutboundWebhook,
} from './hmac.js';

describe('signRestRequest', () => {
  it('matches backend REST HMAC vector', async () => {
    const headers = await signRestRequest({
      secretKey: 'test_secret',
      method: 'POST',
      path: '/api/api-keys',
      body: { name: 'SDK HMAC Test' },
      timestampMs: 1_710_000_000_000,
    });
    assert.equal(
      headers['x-signature'],
      '274c9ff280eadf751530e9e7fce2c2a573d8676b13b7108fe353407df7cc9e00',
    );
  });
});

describe('inboundWebhookCanonicalJson', () => {
  it('strips unknown top-level fields', () => {
    assert.equal(
      inboundWebhookCanonicalJson({ text: 'BUY SOL', extraField: 'ignored' }),
      '{"text":"BUY SOL"}',
    );
  });

  it('matches backend structured buy canonical', () => {
    const body = {
      eventType: 'buy',
      idempotencyKey: 'test-1',
      reasoning: 'test',
      tokenAddress: 'So11111111111111111111111111111111111111112',
      amount: 500,
      unknownField: 'should be stripped',
    };
    assert.equal(
      inboundWebhookCanonicalJson(body),
      '{"amount":500,"eventType":"buy","idempotencyKey":"test-1","reasoning":"test","tokenAddress":"So11111111111111111111111111111111111111112"}',
    );
  });
});

describe('signInboundWebhook', () => {
  it('matches backend inbound vector', async () => {
    const headers = await signInboundWebhook({
      signingSecret: 'test_secret',
      body: {
        text: 'BUY SOL 500 USDC',
        idempotencyKey: 'sdk-test-1',
      },
      timestampSeconds: 1_710_000_000,
    });
    assert.equal(
      headers['X-EZ-Signature'],
      '1d30d896fc609e62bcf9be991c1dc1177a9c63219909869ef2846bad4685de3b',
    );
  });

  it('ignores unknown fields when signing', async () => {
    const body = {
      eventType: 'buy',
      idempotencyKey: 'test-1',
      reasoning: 'test',
      tokenAddress: 'So11111111111111111111111111111111111111112',
      amount: 500,
    };
    const withExtra = { ...body, unknownField: 'strip me' };
    const a = await signInboundWebhook({
      signingSecret: 'secret',
      body,
      timestampSeconds: 1_710_000_000,
    });
    const b = await signInboundWebhook({
      signingSecret: 'secret',
      body: withExtra,
      timestampSeconds: 1_710_000_000,
    });
    assert.equal(a['X-EZ-Signature'], b['X-EZ-Signature']);
    assert.equal(
      a['X-EZ-Signature'],
      'be420f61d91e6b871481774c62f972f0aafa6f5f1a727ba1e4a32558784f77c3',
    );
  });
});

describe('verifyInboundWebhook', () => {
  it('accepts valid signatures', async () => {
    const body = { text: 'BUY SOL 500 USDC', idempotencyKey: 'sdk-test-1' };
    const headers = await signInboundWebhook({
      signingSecret: 'test_secret',
      body,
      timestampSeconds: 1_710_000_000,
    });
    const ok = await verifyInboundWebhook({
      signingSecret: 'test_secret',
      body,
      timestampSeconds: headers['X-EZ-Timestamp'],
      signature: headers['X-EZ-Signature'],
      maxSkewSeconds: 999_999_999,
    });
    assert.equal(ok, true);
  });
});

describe('verifyOutboundWebhook', () => {
  it('verifies signal.execution callbacks', async () => {
    const rawBody = JSON.stringify({
      event: 'signal.execution',
      signalId: 'sig_1',
      developerAgentId: 'agent_1',
      status: 'executed',
      timestamp: '2026-07-06T12:00:00.000Z',
    });
    const timestamp = '2026-07-06T12:00:00.000Z';
    const { createHmac } = await import('node:crypto');
    const signature = createHmac('sha256', 'webhook_secret')
      .update(`${timestamp}.${rawBody}`)
      .digest('hex');

    const ok = await verifyOutboundWebhook({
      secretKey: 'webhook_secret',
      rawBody,
      timestamp,
      signature,
    });
    assert.equal(ok, true);
  });
});
